pkg/noderesourcetopology/cache/overallocating.go

```diff
@@ -0,0 +1,238 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	k8scache "k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+
+	topologyv1alpha1 "github.com/k8stopologyawareschedwg/noderesourcetopology-api/pkg/apis/topology/v1alpha1"
+	listerv1alpha1 "github.com/k8stopologyawareschedwg/noderesourcetopology-api/pkg/generated/listers/topology/v1alpha1"
+
+	"sigs.k8s.io/scheduler-plugins/pkg/noderesourcetopology/stringify"
+
+	"github.com/k8stopologyawareschedwg/podfingerprint"
+)
+
+type NodeIndexer interface {
+	PodNamespacedNamesByNode(logKey, nodeName string) ([]types.NamespacedName, error)
+}
+
+type Overallocating struct {
+	lock             sync.Mutex
+	nrts             nrtStore
+	assumedResources map[string]resourceStore // nodeName -> resourceStore
+	nodeDiscarded    counter
+	lister           listerv1alpha1.NodeResourceTopologyLister
+	indexer          NodeIndexer
+}
+
+func NewOverallocating(lister listerv1alpha1.NodeResourceTopologyLister, indexer NodeIndexer) (*Overallocating, error) {
+	if lister == nil || indexer == nil {
+		return nil, fmt.Errorf("nrtcache: received nil references")
+	}
+
+	nrtObjs, err := lister.List(labels.NewSelector())
+	if err != nil {
+		return nil, err
+	}
+
+	klog.V(3).InfoS("nrtcache: initializing", "objects", len(nrtObjs))
+	return &Overallocating{
+		nrts:             newNrtStore(nrtObjs),
+		assumedResources: make(map[string]resourceStore),
+		nodeDiscarded:    make(counter),
+		lister:           lister,
+		indexer:          indexer,
+	}, nil
+}
+
+func (ov *Overallocating) GetByNode(nodeName string, pod *corev1.Pod) *topologyv1alpha1.NodeResourceTopology {
+	logKey := keyFromPod(pod)
+	ov.lock.Lock()
+	defer ov.lock.Unlock()
+	nrt := ov.nrts.GetByNode(nodeName)
+	if nrt == nil {
+		return nil
+	}
+	nodeAssumedResources := ov.assumedResources[nodeName]
+
+	klog.V(6).InfoS("nrtcache NRT", "logKey", logKey, "vanilla", stringify.NodeResourceTopologyResources(nrt))
+	nodeAssumedResources.UpdateNRT(logKey, nrt)
```

Comment:
I was not aware of this `PostFilter`, so I'll move the logic to over-assume in `Reserve`.
The reason we do the computation here is because we keep a pristine copy of NRT data and we track the overreserve separately. When the client code wants to know the expected availability, we merge the two informations in the returned NRT data.

Reply:
(rewriting the previous answer which was written too hastily). What we do here is:
1. take a deepcopy of the latest NRT data we have (the latest expected we observed, so the one on which the observed set of pods matched the expected set of pods)
2. update **the local deepcopy** with the requests of the pods, creating the overreserve
3. return that data

so the cached copy is not going to be updated here, but only the local copy. This is safe to be done concurrently.
We track the extra resource to overreserve indeed in `Reserve`.

Reply:
ok, I see. Can we rename `GetByNode` to `GetCachedNRTCopy`, and add a comment like (tweak any words if inaccurate):

```
GetCachedNRTCopy retrieves a NRT copy from cache, and then deducts over-reserved resources if necessary.
It will be used as the source of truth across the Pod's scheduling cycle.
```

Reply:
Sure, done.